changes/2020-06-17_encrypt-max-plaintext-length-input/change.md

```diff
@@ -0,0 +1,182 @@
+
+The purpose of this change is to document the correct behavior of this input and
+rename this input to clarify its intent.
+
+This change renames this input within the spec because
+the intention behind the value of this input is that
+it is a bound on the plaintext length,
+not the exact value.
+This is what the Encrypt operation enforces when this value is set,
+and this is how CMMs should interpret this value in GetEncryptionMaterials if it is set.
+Implementations already treat this input with this intent in mind,
+so we should describe it in the specification as such.
+
+This change specifies that the Encrypt operation MUST pass the real length of the plaintext
+to the CMM if known instead of a user supplied bound because the former is more accurate.
+There is no case where a user would benefit from passing
+a less accurate `max plaintext length` to the CMM.
+If there were such a case,
+then such a CMM is not correctly using the intent behind `maxPlaintextLength`.
+It is better to restrict flexibilty here so that users have less opportunity
+to depend on behaviors from CMMs that are misusing this value.
+
+## Drawbacks
+
+This change does not prescribe the name of this input in implementations,
+and thus there will not be a standardized name or way of specifying this input across implementations.
+This is because this control already exists in implementations under different names,
+and it is not worth breaking all customers to rename this input
+without significantly changing this input to benefit the customer.
+
+## Security Implications
+
+This change SHOULD NOT have any security implications.
+
+## Operational Implications
+
+This change will break any JS user that is currently using Encrypt with a known-length plaintext
+and inputting a plaintextLength less than that length.
+
+This change will break any Python user that depends on the source_length being sent to the
+CMM instead of the true length of a known-length plaintext.
+
+## Guide-level Explanation
+
+This change renames the `plaintext length` Encryption input to `max plaintext length` in the specification,
+and clarifies its behavior in the Encrypt operation.
+
+If during the Encrypt operation it is determined that
+the actual plaintext length is greater than a supplied `max plaintext length`,
+the operation MUST fail.
+
+Additionally, when calling the CMM's GetEncryptionMaterials call during the Encrypt operation,
+the operation MUST pass in the length of the input plaintext if it can be known.
+If it can't be known, the operation passes in `max plaintext input` if supplied.
+If it can't be known and no `max plaintext input` is supplied, the operation supplies no value for
+this field.
+
+This change also renames the `plaintext length` field in the CMM interface to `max plaintext length`.
+
+Finally, this change specifies when this input should be available for users to specify:
+if and only if the input plaintext can be of unknown-length.
```

Comment:
How should this be accomplished?
The JS encrypt function takes multiple options.
JavaScript does not have any function overloads.

Should there be multiple `encrypt` functions?
What should they be called?
Or is what it is doing currently adequate?

What does "available" mean?

Reply:
I think answers to these questions are more clearly addressed in the "Reference-level Explanation" section. I was actually a bit unsure what detail to go into in this one. As it is right now, this sentence is not intending to specify what the exact shape of the API should look like, only that if you provide an API that "only" takes in known-length plaintexts, then it doesn't make sense for users to be able to specify this input.